### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by vrom911**

### CODE REVIEW

File: src/PgNamed.hs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+       ) where
+
+import Control.Monad.Except (MonadError (throwError))
+import Data.Bifunctor (bimap)
+import Data.ByteString (ByteString)
+import Data.Char (isAlphaNum)
+import Data.List (lookup)
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.Text (Text)
+import Data.Text.Encoding (decodeUtf8)
```
Comment (chshersh) [inline src/PgNamed.hs:39]: Heh, I think it was painful to write all these imports without `relude`...

File: src/PgNamed.hs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+-- | Wrapper over name of the argument.
+newtype Name = Name
+    { unName :: Text
+    } deriving newtype (Show, Eq, Ord, IsString)
+
+-- | Data type to represent each named parameter.
+data NamedParam = NamedParam
+    { namedParamName  :: Name
+    , namedParamParam :: PG.Action
```
Comment (chshersh) [inline src/PgNamed.hs]: What do you think about making these two fields strict?

File: src/PgNamed.hs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+data PgNamedError
+    -- | Named parameter is not specified.
+    = PgNamedParam Name
+    -- | Query has no names inside but was called with named functions,
+    | PgNoNames PG.Query
+    -- | Query contains an empty name.
+    | PgEmptyName PG.Query
+
+
+-- | Type alias for 'PgNamedError'.
```
Comment (chshersh) [inline src/PgNamed.hs]: Just to be a little bit more specific
```suggestion
-- | Type alias for monads that can throw errors of the 'PgNamedError' type.
```

File: src/PgNamed.hs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    isNameChar :: Char -> Bool
+    isNameChar c = isAlphaNum c || c == '_'
+
+
+-- | Returns the list of values to use in query by given list of 'Name's.
+namesToRow
+    :: forall m . WithError m
+    => NonEmpty Name  -- ^ List of the names used in query
+    -> [NamedParam]   -- ^ List of the named parameters
+    -> m (NonEmpty PG.Action)
```
Comment (chshersh) [inline src/PgNamed.hs:127]: It's nice that this function is abstracted over `WithError` monad. This means that in this package we can also implement `queryNamed` and `executeNamed` functions 👍 No necessary in this PR, but should be possible in theory.

File: postgresql-simple-named.cabal [hunk 5]
```diff
@@ -1,8 +1,22 @@
...
-description:         Implementation of named parameters for `postgresql-simple` library
+description:
+    Implementation of named parameters for `postgresql-simple` library.
+    .
+    Here is an exaple of how it could be used in your code:
+    .
+    > queryNamed [sql|
+    >     SELECT *
+    >     FROM table
+    >     WHERE foo = ?foo
```
Comment (chshersh) [inline postgresql-simple-named.cabal:13]: Wow, it's very nice to such docs in the `.cabal` file description! Good idea!

**Review by chshersh — COMMENTED** [review]
Very nice implementation!

File: src/PgNamed.hs (hunk 4 above)
Comment (vrom911) [reply src/PgNamed.hs:127]: I made an issue: #3

**Review by chshersh — APPROVED** [review]

**Pull request merged**